Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -1119,11 +1202,165 @@ ErrorOr<u32> TagTree::read_value(u32 x, u32 y, Function<ErrorOr<bool>()> const&
+
+        dbgln_if(JPEG2000_DEBUG, "Sub-band rect: {}", sub_band.rect);
+
+        // Compute tile size at resolution level r.
+        auto ll_rect = context.siz.reference_grid_coordinates_for_ll_band(tile.rect, component_index, r, N_L);
+
+        // B.6
+        // (B-16)
+        int num_precincts_wide = 0;
+        int num_precincts_high = 0;
+        int PPx = coding_parameters.precinct_sizes[r].PPx;
+        int PPy = coding_parameters.precinct_sizes[r].PPy;
+
+        if (ll_rect.width() != 0)
+            num_precincts_wide = ceil_div(ll_rect.right(), 1 << PPx) - (ll_rect.left() / (1 << PPx));
+        if (ll_rect.height() != 0)
+            num_precincts_high = ceil_div(ll_rect.bottom(), 1 << PPy) - (ll_rect.top() / (1 << PPy));
+
+        sub_band.num_precincts_wide = num_precincts_wide;
+        sub_band.num_precincts_high = num_precincts_high;
+
+        auto precinct_origin = IntPoint { ll_rect.x() & ~((1 << PPx) - 1), ll_rect.y() & ~((1 << PPy) - 1) };
+
+        if (r > 0) {
+            PPx--;
+            PPy--;
+            precinct_origin /= 2;
+        }
+
+        // B.7
+        // (B-17)
+        // (The r > 0 check was done right above already.)
+        int xcb_prime = min(coding_parameters.code_block_width_exponent, PPx);
+
+        // (B-18)
+        // (The r > 0 check was done right above already.)
+        int ycb_prime = min(coding_parameters.code_block_height_exponent, PPy);
+
+        for (int precinct_y_index = 0; precinct_y_index < num_precincts_high; ++precinct_y_index) {
+            for (int precinct_x_index = 0; precinct_x_index < num_precincts_wide; ++precinct_x_index) {
+                auto precinct_rect = IntRect({ precinct_x_index * (1 << PPx), precinct_y_index * (1 << PPy), 1 << PPx, 1 << PPy });
+                precinct_rect.set_location(precinct_rect.location() + precinct_origin);
+
+                sub_band.precincts.append(TRY(make_precinct(sub_band, precinct_rect, xcb_prime, ycb_prime)));
+            }
+        }
+
+        return {};
+    };
+
+    auto make_component = [&](TileData const& tile, int component_index) -> ErrorOr<DecodedTileComponent> {
+        DecodedTileComponent component;
+        component.rect = context.siz.reference_grid_coordinates_for_tile_component(tile.rect, component_index);
+
+        dbgln_if(JPEG2000_DEBUG, "making nLL for component {}", component_index);
+        TRY(make_sub_band(tile, component_index, component.nLL, JPEG2000::SubBand::HorizontalLowpassVerticalLowpass, 0));
+
+        auto N_L = context.coding_style_parameters_for_component(tile, component_index).number_of_decomposition_levels;
+        for (int resolution_level = 1; resolution_level <= N_L; ++resolution_level) {
+            DecodedTileComponent::DecodedSubBands sub_bands;
+            for (auto const& [sub_band_index, sub_band] : enumerate(DecodedTileComponent::SubBandOrder)) {
```

Comment:
why the `const&`, shouldnt all parts of the `*iterator` be trivially copyable (an int and an enum)

Reply:
Yes, this was silly, thanks.

(I think with `auto` the index is no longer a reference, but the `sub_band` still is (but the compiler should be able to figure out it can copy it instead.))